# BLIBIOTECAS

In [4]:
from pathlib import Path
import progressbar as pb # Visualizar barra de progresso, bom para operações de longa duração
import pandas as pd
import numpy as np
import random
import warnings # supress warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

from sklearn.metrics.pairwise import cosine_similarity
# Add some convenience functions to Pandas DataFrame.
pd.options.display.max_rows = 10
pd.options.display.float_format = '{:.3f}'.format

# Exercício Prático - Matrizes Esparsas

# 1. Carregar os datasets disponibilizados pela DNC ou fazer download via Kaggle dos datasets 'arashnic/book-recommendation-dataset'

In [5]:
%env KAGGLE_USERNAME = tomasmenegatto
%env KAGGLE_KEY = 11458e6485cef3e20517783268b9afac

!content/book-recommendation-dataset
!kaggle datasets download -d arashnic/book-recommendation-dataset --unzip -p content/book-recommendation-dataset
!ls content/book-recommendation-dataset

DATA_PATH = Path("content/book-recommendation-dataset")

book_df = pd.read_csv(DATA_PATH / 'Books.csv')
ratings_df = pd.read_csv(DATA_PATH / 'Ratings.csv')
users_df = pd.read_csv(DATA_PATH / 'Users.csv')

env: KAGGLE_USERNAME=tomasmenegatto
env: KAGGLE_KEY=11458e6485cef3e20517783268b9afac
/bin/bash: content/book-recommendation-dataset: Is a directory
 38% 9.00M/23.8M [00:00<00:00, 45.7MB/s]
100% 23.8M/23.8M [00:00<00:00, 79.0MB/s]
Books.csv  Ratings.csv	Users.csv


# 2. Amostrar e obter a matriz de relacionamento entre os clientes e os livros.


##2.1. Matriz de Similaridade

In [6]:
# Faz uma cópia para que o dataset original não seja alterado
def start_pipeline(dataf):
    return dataf.copy()

# Agrupa o dateset por usuários e conta quantos livros diferentes cada um tem
# Ordena para que os usuários com mais livros fiquem no topo do dataset
def get_user_counts(dataf): 
  return dataf.groupby(by='User-ID', as_index=False) \
    .agg({'ISBN': pd.Series.nunique}) \
    .rename(columns={'ISBN': 'ISBN_count'}) \
    .set_index('User-ID') \
    .sort_values('ISBN_count', ascending = False)

# Agrupa o dateset por livros e conta quantos usuários diferentes cada lirvro tem
# Ordena para que os livros com mais usuários diferentes fiquem no topo do dataset
def get_book_counts(dataf): 
  return dataf.groupby(by='ISBN', as_index=False) \
    .agg({'User-ID': pd.Series.nunique}) \
    .rename(columns={'User-ID': 'User-ID_count'}) \
    .set_index('ISBN') \
    .sort_values('User-ID_count', ascending = False)

#Recebe o dataset e o tamanho da amostra
 # Retorna as n primeiras linhas do dataset recebido em formato de lista
def get_n_top_values(dataf, n): 
  return dataf.head(sample_size) \
    .index \
    .tolist()

 # Executa as funções acima para rankear os usuários
sample_size = 1000
top_user_ids = ratings_df \
  .pipe(start_pipeline) \
  .pipe(get_user_counts) \
  .pipe(get_n_top_values, sample_size)

# Executa as funções acima para rankear os livros
sample_size = 999  
top_book_ids = ratings_df \
  .pipe(start_pipeline) \
  .pipe(get_book_counts) \
  .pipe(get_n_top_values, sample_size)

##2.2. Matriz de relacionamento entre clientes

In [7]:
#Transforma a amostra em um dataset, renomeando as colunas.
def transform_sample(dataf):
  return pd.DataFrame({
      "user": dataf["User-ID"],
      "book": dataf["ISBN"],
      "action": dataf["Book-Rating"]
  })

# Seleciona só os User-IDs que estão entre os top Users, transforma amostra em um dataframe e organiza o dataset em uma matriz Cliente x Produto 
matriz_de_relacionamento_user_sampled = ratings_df[ratings_df["User-ID"].isin(top_user_ids)] \
  .pipe(transform_sample) \
  .pivot_table(index="user", columns="book", values="action").fillna(0)

In [8]:
# Verifica a dimensão da matriz C x P - 1000 Clientes (tamanho da amostra escolhida) e 212.171 Produtos (livros)
matriz_de_relacionamento_user_sampled.shape

(1000, 212171)

In [9]:
matriz_de_relacionamento_user_sampled.head(10)

book   0330299891   0375404120   0586045007  ...  cn108465  cn113107  Ô½crosoft
user                                         ...                               
254         0.000        0.000        0.000  ...     0.000     0.000      0.000
2033        0.000        0.000        0.000  ...     0.000     0.000      0.000
2276        0.000        0.000        0.000  ...     0.000     0.000      0.000
2766        0.000        0.000        0.000  ...     0.000     0.000      0.000
2977        0.000        0.000        0.000  ...     0.000     0.000      0.000
3363        0.000        0.000        0.000  ...     0.000     0.000      0.000
3757        0.000        0.000        0.000  ...     0.000     0.000      0.000
4017        0.000        0.000        0.000  ...     0.000     0.000      0.000
4385        0.000        0.000        0.000  ...     0.000     0.000      0.000
6242        0.000        0.000        0.000  ...     0.000     0.000      0.000

[10 rows x 212171 columns]

##2.3. Matriz de relacionamento entre os livros

In [10]:
# Seleciona só os Books que estão entre os top books, transforma a amostra em um dataframe e organiza o dataset em uma matriz Cliente x Produto 
matriz_de_relacionamento_book_sampled = ratings_df[ratings_df["ISBN"].isin(top_book_ids)] \
  .pipe(transform_sample) \
  .pivot_table(index="user", columns="book", values="action").fillna(0)

In [11]:
# Verifica as dimensões da matriz CxP transposta, ou seja: PxC
matriz_de_relacionamento_book_sampled.T.shape

(999, 35612)

In [12]:
matriz_de_relacionamento_book_sampled.T.head(10)

user        9       14      16      17      ...  278843  278844  278851  278854
book                                        ...                                
000649840X   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
002542730X   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
0060008032   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
0060096195   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
006016848X   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
0060173289   0.000   0.000   0.000   0.000  ...   9.000   0.000   0.000   0.000
0060175400   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
0060188731   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
006019491X   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000
0060199652   0.000   0.000   0.000   0.000  ...   0.000   0.000   0.000   0.000

[10 rows x 35612 columns]

# 3. Obter a matriz de relacionamento ESPARSA entre os clientes e os livros

In [13]:
# Para obter a matriz esparsa basta dummificar as colunas User-ID e Books do dataset originial

matriz_relacionamento_esparsa = pd.get_dummies(ratings_df.sample(1000), columns=["User-ID", "ISBN"], sparse=True)

matriz_relacionamento_esparsa

# Se tentássemos fazer o mesmo para matriz não esparsa, não teríamos memória suficiente

Book-Rating  User-ID_183  ...  ISBN_9724722678  ISBN_9788479790738
902471             0            0  ...                0                   0
1025125            0            0  ...                0                   0
434689             0            0  ...                0                   0
453573             0            0  ...                0                   0
1112797            0            0  ...                0                   0
...              ...          ...  ...              ...                 ...
836488             0            0  ...                0                   0
401203             0            0  ...                0                   0
218755             0            0  ...                0                   0
812814             9            0  ...                0                   0
109685             0            0  ...                0                   0

[1000 rows x 1837 columns]

#4. Obter a matriz de similaridade (via cosine_similarity) entre livros com a matriz NÃO ESPARSA.

##4.1. Para Clientes

In [14]:
# Calculando a similaridade entre os usuários através da multiplicação da matriz CxP pela sua transposta
cos_user = cosine_similarity(matriz_de_relacionamento_user_sampled)
#Imprimindo o resultado
print(cos_user)

[[1.         0.0528255  0.         ... 0.         0.         0.        ]
 [0.0528255  1.         0.04675324 ... 0.         0.         0.        ]
 [0.         0.04675324 1.         ... 0.02334216 0.         0.        ]
 ...
 [0.         0.         0.02334216 ... 1.         0.         0.        ]
 [0.         0.         0.         ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.         0.         1.        ]]


##4.2. Para Livros

In [15]:
# Calculando a similaridade entre os livros através da multiplicação da matriz Produto x Cliente pela sua transposta 
cos_item = cosine_similarity(matriz_de_relacionamento_book_sampled.T)
# Imprimindo o resultado
print(cos_item)

[[1.         0.05983705 0.02857749 ... 0.02202635 0.01551346 0.        ]
 [0.05983705 1.         0.01555774 ... 0.02620312 0.         0.        ]
 [0.02857749 0.01555774 1.         ... 0.04822782 0.         0.        ]
 ...
 [0.02202635 0.02620312 0.04822782 ... 1.         0.         0.        ]
 [0.01551346 0.         0.         ... 0.         1.         0.03301185]
 [0.         0.         0.         ... 0.         0.03301185 1.        ]]


#5. Obter a matriz de similaridade (via multiplicação de matrizes) entre livros com a matriz ESPARSA.

In [19]:
matriz_similaridade_esparsa = matriz_relacionamento_esparsa.T.multiply(matriz_relacionamento_esparsa)
matriz_similaridade_esparsa

1831  3963  4165  ...  User-ID_99227  User-ID_9992  User-ID_99951
1831            nan   nan   nan  ...            nan           nan            nan
3963            nan   nan   nan  ...            nan           nan            nan
4165            nan   nan   nan  ...            nan           nan            nan
6206            nan   nan   nan  ...            nan           nan            nan
6879            nan   nan   nan  ...            nan           nan            nan
...             ...   ...   ...  ...            ...           ...            ...
User-ID_98741   nan   nan   nan  ...            nan           nan            nan
User-ID_99032   nan   nan   nan  ...            nan           nan            nan
User-ID_99227   nan   nan   nan  ...            nan           nan            nan
User-ID_9992    nan   nan   nan  ...            nan           nan            nan
User-ID_99951   nan   nan   nan  ...            nan           nan            nan

[2837 rows x 2837 columns]

#6. Qual a diferença de tamanhos das matrizes?

In [20]:
# Criando função para mensurar a quantidade de memória utilizada
def print_memory_usage_of_data_frame(df, bytes_to_mb_div = 0.000001):
    mem = round(df.memory_usage().sum() * bytes_to_mb_div, 3) 
    print("Memory usage is " + str(mem) + " MB")

In [21]:
# Exibindo o tempo de processamento
%%time

print_memory_usage_of_data_frame(matriz_de_relacionamento_book_sampled)

Memory usage is 286.207 MB
CPU times: user 10.9 ms, sys: 0 ns, total: 10.9 ms
Wall time: 12.5 ms


In [22]:
# Exibindo o tempo de processamento
%%time

print_memory_usage_of_data_frame(matriz_similaridade_esparsa)

Memory usage is 64.411 MB
CPU times: user 89.1 ms, sys: 2.75 ms, total: 91.9 ms
Wall time: 92.9 ms


#**RESPOSTA:** Enquanto a matriz densa ocupa 286.207 MB, a matriz esparsa tem apenas 64.411 MB (22% da matriz densa)